Add TRT-LLM Gen. AI Autoscaling & Load Balancing Guide (#95)
triton-inference-server/tutorials: Deployment/Kubernetes/TensorRT-LLM_Autoscaling_and_Load_Balancing/README.md

  * [Deploying Models Too Large for a Single GPU](#deploying-models-too-large-for-a-single-gpu)
  * [Utilizing Multiple GPU SKUs](#utilizing-multiple-gpu-skus)
  * [Monitoring Triton in Kubernetes](#monitoring-triton-in-kubernetes)
* [Developing this Guide](#developing-this-guide)

Prior to beginning this guide/tutorial you will need a couple of things.

* Kubernetes Control CLI (`kubectl`)
  [ [documentation](https://kubernetes.io/docs/reference/kubectl/introduction/)
  | [download](https://kubernetes.io/releases/download/) ]
* Helm CLI (`helm`)
  [ [documentation](https://helm.sh/)
  | [download](https://helm.sh/docs/intro/install) ]
* Docker CLI (`docker`)
  [ [documentation](https://docs.docker.com/)
  | [download](https://docs.docker.com/get-docker/) ]
* Decent text editing software for editing YAML files.
* Kubernetes cluster.
* Fully configured `kubectl` with administrator permissions to the cluster.

## Cluster Setup

The following instructions are setting up Horizontal Pod Autoscaling (HPA) for Triton Server in a Kubernetes cluster.

### Prerequisites

This guide assumes that all nodes with NVIDIA GPUs have the following:

> - A node label of `nvidia.com/gpu=present` to more easily identify nodes with NVIDIA GPUs.

Author: the point wasn to not soley rely on the gpu-feature-discovery service. thus allow us control over where the NVIDIA services interact because a few of them fail when no GPU is present.

> - A node label of `nvidia.com/gpu=present` to more easily identify nodes with NVIDIA GPUs.

Reviewer: It does seem like we'd want to proscribe an automated  way to handle this - is there a way to use either the label added by gpu feature discovery - or the manual added label? - can be for the future.

- A node taint of `nvidia.com/gpu=present:NoSchedule` to prevent non-GPU pods from being deployed to GPU nodes.

> When using a Kubernetes provider like AKS, EKA, or GKE, it is usually best to use their interface when configuring nodes
> instead of using `kubectl` to do it directly.

> ### Core Cluster Services

Author: It doesn't work on standard EKS node types. Unfortunately, it requires an Ubuntu based node type which is more complex / difficult to setup and potentially increases costs.

2.  Run the command below to install the service.

    ```bash
    helm install -n kube-system node-feature-discovery kube-nfd/node-feature-discovery \
      --set nameOverride=node-feature-discovery \
      --set worker.tolerations[0].key=nvidia.com/gpu \
      --set worker.tolerations[0].operator=Exists \
      --set worker.tolerations[0].effect=NoSchedule
    ```

#### NVIDIA Device Plugin for Kubernetes

1.  This step is unnecessary if the Device Plugin has already been installed in your cluster.
    Cloud provider turnkey Kubernetes clusters, such as those from AKS, EKS, and GKE, often have the Device Plugin
    automatically once a GPU node as been added to the cluster.

    To check if your cluster requires the NVIDIA Device Plugin for Kubernetes, run the following command and inspect
    the output for `nvidia-device-plugin-daemonset`.

    ```bash
    kubectl get daemonsets -n kube-system
    ```

    Example output:
    ```text
    NAME                             DESIRED  CURRENT  READY  UP-TO-DATE  AVAILABLE
    kube-proxy                       6        6        6      6           6
    nvidia-device-plugin-daemonset   6        6        6      6           6
    ```

> 2.  Run the command below to install the plugin.

Author: fixed up.

    Once installed it will provide containers access to GPUs in your clusters.

    For additional information, see
    [Github/NVIDIA/k8s-device-plugin](https://github.com/NVIDIA/k8s-device-plugin/blob/main/README.md).

    ```bash
    kubectl create -f https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/v0.15.0/deployments/static/nvidia-device-plugin.yml
    ```

#### NVIDIA GPU Feature Discovery Service

> 1.  Use the YAML contents below create a file named `nvidia_gpu-feature-discovery_daemonset.yaml`.

Author: Fixed up.

    [GitHub/NVIDIA](https://raw.githubusercontent.com/NVIDIA/gpu-feature-discovery/v0.8.2/deployments/static/gpu-feature-discovery-daemonset.yaml)
    and modified specifically for Triton Server autoscaling.

    ```bash
    curl https://raw.githubusercontent.com/NVIDIA/gpu-feature-discovery/v0.8.2/deployments/static/gpu-feature-discovery-daemonset.yaml \
      >  nvidia_gpu-feature-discovery_daemonset.yaml
    ```

2.  Then run the command below to install the

    ```bash
    kubectl apply -f ./nvidia_gpu-feature-discovery_daemonset.yaml
    ```

### Metrics Collection Services

Your cluster is now up, running, and can even assign GPU resources to containers.
Next, we have to setup metrics collection for DCGM and Triton Server.
Metrics provide insight to the Kubernetes Horizontal Pod Autoscaler service and enable it to make autoscaling decisions based
on the utilization and availability of deployed models.

#### Create a Monitoring Namespace

Create the `monitoring` namespace in your cluster for all of the metrics and monitoring services.

1.  Run the command below to create the namespace.

    ```bash
    kubectl create namespace monitoring
    ```

#### Prometheus Services

> We a service to collect, store, aggregate, and provide metrics collected from your cluster and the services deployed in it.

Author: Fixed up. (the word "need" was needed 🤦‍♂️)

One of the easiest ways to do this is to leverage the functionality of the [Prometheus Metrics Server](https://prometheus.io/).

> Using the following steps, we'll install the Prometheus Stack for Kubernetes Helm chart so that we can leverage Prometheus.

Author: Fixed up.

1.  Add the Prometheus Community chart repository to the local cache.

    ```bash
    helm repo add prometheus-community https://prometheus-community.github.io/helm-charts \
      && helm repo update
    ```

2.  Run the command below to install the Prometheus Kubernetes Stack Helm chart.

    ```bash
    helm install -n monitoring prometheus prometheus-community/kube-prometheus-stack \
      --set tolerations[0].key=nvidia.com/gpu \
      --set tolerations[0].operator=Exists \
      --set tolerations[0].effect=NoSchedule
    ```

#### NVIDIA Data Center GPU Manager (DCGM) Exporter

The best solution for management of GPUs in your cluster is
[NVIDIA DCGM](https://docs.nvidia.com/data-center-gpu-manager-dcgm)(DCGM).
However, for this example we do not need the entirety of the DCGM stack.

> Instead, we'll use the steps below to install the [DCGM Exporter](https://github.com/NVIDIA/dcgm-exporter) to enable the

Author: Fixed up.

collection of GPU metrics in your cluster.

1.  Add the NVIDIA DCGM chart repository to the local cache.

    ```bash
    helm repo add nvidia-dcgm https://nvidia.github.io/dcgm-exporter/helm-charts \
      && helm repo update
    ```

2.  Use the YAML contents below to create a file named `nvidia_dcgm-exporter_values.yaml`.

>     > [nvidia_dcgm-exporter_values.yaml](nvidia_dcgm-exporter_values.yaml)

Author: Good point, changed the wording for better clarity.

    The contents above were generated using `helm show values nvidia-dcgm/dcgm-exporter` and then modified specifically for
    Triton Server autoscaling.

4.  Install the DCGM Exporter Helm chart using the following command.

    ```bash
    helm install -n monitoring dcgm-exporter nvidia-dcgm/dcgm-exporter --values nvidia_dcgm-exporter_values.yaml
    ```

#### Connect DCGM and Triton Metrics to Prometheus

We need to provide a mechanism that will scrape the metrics produced by DCGM Exporter and inject them into the Prometheus
metrics server.
The steps below will setup a Prometheus Adapter that collects metrics from every DCGM Exporter worker and provides them to

> Prometheus.

Reviewer 2: You're right, not the Kubernetes metrics server -- but the adapter makes it so that you can query Prometheus through the Kubernetes Metrics API (which is what the HPA uses). It doesn't affect what DCGM provides to Prometheus, only what Prometheus provides to the HPA

      --set additionalLabels.release=prometheus
    ```

2.  To verify that the adapter is installed and configured correctly, wait for at least 60 seconds and then run the
    following command.
    It is important to note that here is a noticeable delay between the adapter being installed and the availability of custom
    metrics.

    ```bash
    kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1
    ```

    If the command fails, wait longer and retry. If the command fails for more than a few minutes then the adapter is
    misconfigured and will require intervention.

#### Triton Metrics Prometheus Rule

Prometheus rules provide a mechanism for generation of metrics data using a formula that operates on data being collected by
Prometheus.
We'll create a set of rules specific to Triton Server which generate metrics useful for autoscaling.

1.  Use the YAML contents below to create a file named `triton-metrics_prometheus-rule.yaml`.

    > [triton-metrics_prometheus-rule.yaml](triton-metrics_prometheus-rule.yaml)

2.  Run the following command to create the necessary Prometheus Rule in the cluster. _Note_ that the rule will be created

>     in the namespace of your current context, usually `default`.

Author: Yep. Been testing it like this for weeks now.

    If you prefer to install it in a different namespace you can either update your context or add `-n <desired_namespace>` to
    the command.

    ```bash
    kubectl apply -f ./triton-metrics_prometheus-rule.yaml
    ```

In all of value files for the example Helm chart, the horizontal-pod autoscaler is configured to use the

> `triton:queue_compute:ratio` metric provide by the above rules.

Author: Fixed up.

The benefit of using this metric is that it is hardware and model independent since it measures the ratio between the time a
request spends in the inference queue to the time it takes to complete once it has left the queue.

> This kind of metric allows he performance of models on diverse hardware to compared to each other.

Author: Fixed up.

pods deployed on the same node.
This enables the generation script to detect that the plan and engine generation steps have been completed and not repeat work.

When Triton Server is started, the same host folders will be mounted to its container and Triton will use the pre-generated
model plan and engine files.
This drastically reduces the time required for subsequent pod starts on the same node.

1.  Create a Python file with the content below named `server.py`.

    > [server.py](containers/server.py)

    This solution could be further improved by adding a network storage location shared by all nodes in a cluster that
    could be used to globally cache per model/GPU plan and engine files.
    Subsequent pod starts on new nodes with the same GPU could download the pregenerated files instead generating them
    locally.
    This could save significant sime depending on the delta between the time to download the files instead of generating them
    (likely several seconds at least).

#### Custom Container Image

1.  Create a container file with the content below named `triton_trt-llm.containerfile`.

>     > [triton_trt-llm.containerfile](containers/server.containerfile)

Author: Changed.

2.  Run the following command to create a custom Triton Inference Server w/ all necessary tools to generate TensorRT-LLM
    plan and engine files. In this example we'll use the tag `24.04` to match the date portion of `24.04-trtllm-python-py3`
    from the base image.

    ```bash
    docker build \
      --file ./triton_trt-llm.containerfile \
      --rm \
      --tag triton_trt-llm:24.04 \
      .
    ```

    ##### Custom Version of Triton CLI

    This custom Triton Server container image makes use of a custom version of the Triton CLI.
    The relevant changes have been made available as a

>     [topic branch](https://github.com/triton-inference-server/triton_cli/tree/jwyman/aslb-mn) in the Triton CLI repository on

Reviewer 2: Nevermind, saw the note about this lower down

    In this example, we'll use the `nvcr.io/nvaie/staging` repository for demonstration purposes.
    You will need to determine which repositories you have write access to that your cluster can also access.

    1. First, re-tag the container image with the repository's name like below.

        ```bash
        docker tag \
          triton_trt-llm:24.04 \
          nvcr.io/nvaie/staging/triton_trt-llm:24.04
        ```

    2. Next, upload the container image to your repository.

        ```bash
        docker push nvcr.io/nvaie/staging/triton_trt-llm:24.04
        ```

#### Kubernetes Pull Secrets

If your container image repository requires credentials to download images from, then you will need to create a Kubernetes
docker-registry secret.
We'll be using the `nvcr.io` container image repository example above for demonstration purposes.
Be sure to properly escape any special characters such as `$` in the password or username values.

1.  Use the command below to create the necessary secret.  Secrets for your repository should be similar, but not be identical
to the example below.

    ```bash
    kubectl create secret docker-registry ngc-container-pull \

>       --docker-password='dGhpcyBpcyBub3QgYSByZWFsIHNlY3JldC4gaXQgaXMgb25seSBmb3IgZGVtb25zdHJhdGlvbiBwdXJwb3Nlcy4' \

Author: That is a placeholder.

```
> base64 decode 'dGhpcyBpcyBub3QgYSByZWFsIHNlY3JldC4gaXQgaXMgb25seSBmb3IgZGVtb25zdHJhdGlvbiBwdXJwb3Nlcy4='
this is not a real secret. it is only for demonstration purposes.

>       --docker-server='nvcr.io'

Author: Fixed up.

    You can verify that the secret was created correctly using the following command and inspecting its output for the secret
    you're looking for.

    ```bash
    kubectl get secrets
    ```

3.  Ensure the contents of the secret are correct, you can run the following command.

    ```bash
    kubectl get secret/ngc-container-pull -o yaml
    ```

    You should see an output similar to the following.

    ```yaml
    apiVersion: v1
    data:
      .dockerconfigjson: eyJhdXRocyI6eyJudmNyLmlvIjp7InVzZXJuYW1lIjoiJG9hdXRodG9rZW4iLCJwYXNzd29yZCI6IlZHaHBjeUJwY3lCdWIzUWdZU0J5WldGc0lITmxZM0psZEN3Z2FYUWdhWE1nYjI1c2VTQm1iM0lnWkdWdGIyNXpkSEpoZEdsdmJpQndkWEp3YjNObGN5ND0iLCJhdXRoIjoiSkc5aGRYUm9kRzlyWlc0NlZrZG9jR041UW5CamVVSjFZak5SWjFsVFFubGFWMFp6U1VoT2JGa3pTbXhrUTNkbllWaFJaMkZZVFdkaU1qVnpaVk5DYldJelNXZGFSMVowWWpJMWVtUklTbWhrUjJ4MlltbENkMlJZU25kaU0wNXNZM2swWjFWSGVHeFpXRTVzU1VjMWJHUnRWbmxKU0ZaNldsTkNRMWxZVG14T2FsRm5aRWM0WjJGSGJHdGFVMEo1V2xkR2MwbElUbXhaTTBwc1pFaE5hQT09In19fQ==
    kind: Secret
    metadata:
      name: ngc-container-pull
      namespace: default
    type: kubernetes.io/dockerconfigjson
    ```

>     The value of `.dockerconfigjson` is a base-64 encoded string which can be decoded into the following.

Author: Added.

    ```json
    {
      "auths": {
        "nvcr.io": {
          "username":"$oauthtoken",
          "password":"VGhpcyBpcyBub3QgYSByZWFsIHNlY3JldCwgaXQgaXMgb25seSBmb3IgZGVtb25zdHJhdGlvbiBwdXJwb3Nlcy4gUGxlYXNlIG5ldmVyIHVzZSBCYXNlNjQgdG8gaGlkZSByZWFsIHNlY3JldHMh",
          "auth":"JG9hdXRodG9rZW46VkdocGN5QnBjeUJ1YjNRZ1lTQnlaV0ZzSUhObFkzSmxkQ3dnYVhRZ2FYTWdiMjVzZVNCbWIzSWdaR1Z0YjI1emRISmhkR2x2YmlCd2RYSndiM05sY3k0Z1VHeGxZWE5sSUc1bGRtVnlJSFZ6WlNCQ1lYTmxOalFnZEc4Z2FHbGtaU0J5WldGc0lITmxZM0psZEhNaA=="
        }
      }
    }
    ```

>     The values of `password` and `auth` are also base-64 encoded string.

Author: I'm seeing
```
> base64 decode 'VGhpcyBpcyBub3QgYSByZWFsIHNlY3JldCwgaXQgaXMgb25seSBmb3IgZGVtb25zdHJhdGlvbiBwdXJwb3Nlcy4gUGxlYXNlIG5ldmVyIHVzZSBCYXNlNjQgdG8gaGlkZSByZWFsIHNlY3JldHMh'
This is not a real secret, it is only for demonstration purposes. Please never use Base64 to hide real secrets!
```

Same outcome (different values obviously) for the real secret in the testing cluster.

Wonder why there's a discrepancy? 🤔

    We recommend inspecting the values of the following values:

    * Value of `.auths['nvcr.io'].username`.

>     * Base64 decoded value of `.auths['nvcr.io'].password`.

Author: You want this as a note and not a list? Sorry, a little confused by this comment.

>     * Base64 decoded value of `.auths['nvcr.io'].password`.

Reviewer 2: I meant the base64 decoded thing -- in my cluster only auth was base64 encoded, not password. Also not sure why there's a discrepancy.

    * Base64 decoded value of `.auths['nvcr.io'].auths`.

## Triton Deployment

### Deploying Single GPU Models

Deploying Triton Server with a model that fits on a single GPU is straightforward using the steps below.

1.  Create a custom values file with required values:

    * Container image name.
    * Model name.
    * Supported / available GPU(s).
    * Image pull secrets (if necessary).
    * Hugging Face secret name.

    The provided sample Helm [chart](./chart/) include several example values files such as

>     [llama-3-8b_values.yaml](client/llama-3-8b-instruct_values.yaml).

Author: Fixed up.

    Triton Server.
    Without inference queries, there are no metrics generated and thus the HPA controller reports the metric's current values
    as `<unknown>`.
    This issue should resolve itself once clients begin sending inference queries to Triton Server.

4.  Uninstalling the Chart

    Uninstalling a Helm chart is as straightforward as running the command below.
    This is useful when experimenting with various options and configurations.

    ```bash
    helm uninstall <installation_name>
    ```

### Deploying Models Too Large for a Single GPU

Given the memory requirements of some AI models it is not possible to host them using a single device.
Triton and TensorRT-LLM provide a mechanism to enable a large model to be hosted by multiple GPU devices working in concert.
The provided sample Helm [chart](./chart/) provides a mechanism for taking advantage of this capability.

To enable this feature, adjust the `model.tensorrtLlm.parallelism.tensor` value to an integer greater than 1.
Configuring a model to use tensor parallelism enables the TensorRT-LLM runtime to effectively combine the memory of multiple
GPUs to host a model too large to fit on a single GPU.

> Similarly, changing the value of  `model.tensorrtLlm.parallelism.tensor` will enable pipeline parallelism.

Reviewer: ```suggestion
Similarly, changing the value of  `model.tensorrtLlm.parallelism.pipeline` will enable pipeline parallelism.
```

Author: fixed up

Pipeline parallelism is used to combine the compute capacity of multiple GPUs to process inference requests in parallel.

The number of GPUs required to host the model is equal to product of the values of `.tensor` and `.pipeline`.
It is important to note that the GPUs used to host a model must reside on the same node.

> _Combining GPUs which reside on separate nodes is not covered in this guide._

Author: Fixed up.

### Utilizing Multiple GPU SKUs

Given the relative limited availability of certain SKUs of GPU, it is not uncommon for services to be required to operate on a
mix of GPU hardware.

> For example, the number of nodes with NVIDIA Hopper based devices might be insufficient to me load requirements and your

Author: fixed up.

> clusters have spare nodes with NVIDIA Ampere based devices.

Author: fixed up.

In this scenario, it would make sense to create multiple deployment of the same model using the steps
[above](#deploying-single-gpu-models) and placing them all behind a single Kubernetes service for load-balancing needs.

> Doing so will enable both SKU of devices to automatically scale independently and provide compute capacity for the service.

Author: fixed up.

To achieve this, we can update the chart to not create a service with our deployment and to include the selector labels
specified by the shared service.

> In the example below, we'll assume the service has already been created and its selector is set to `model=llama-3-8b`.

Author: isn't that in the chart already?

> In the example below, we'll assume the service has already been created and its selector is set to `model=llama-3-8b`.

Reviewer 2: Ah ok I didn't actually deploy this part so didn't realize it

```bash
helm install llama-3-8b-a100 ./chart/. \
  --values ./chart/values.yaml \
  --values ./chart/llama-3-8b \
  --set 'triton.image.name=<custom_image_name>' \
  --set 'gpu[0]=NVIDIA-A100-SXM4-40GB' \
  --set 'kubernetes.labels[0].model=llama-3-8b' \
  --set 'kubernetes.noService=true'

helm install llama-3-8b-h100 ./chart/. \
  --values ./chart/values.yaml \
  --values ./chart/llama-3-8b \
  --set 'triton.image.name=<custom_image_name>' \
  --set 'gpu[0]=NVIDIA-H100-SXM5-80GB' \
  --set 'kubernetes.labels[0].model=llama-3-8b' \
  --set 'kubernetes.noService=true'
```

The result will be two deployments in your cluster, both of which are part of your service's load-balancing pool.

```bash
kubectl get deployments --selector='model=llama-3-8b'
NAME                    READY   UP-TO-DATE   AVAILABLE
llama-3-8b-a100         1/1     1            1
llama-3-8b-h100         1/1     1            1
```

### Monitoring Triton in Kubernetes

Monitoring Triton in Kubernetes can be done using the Prometheus software installed as part of the

> [Prometheus Services](#prometheus-services) second of this document.

Author: fixed up.

Without the provided labels, it would not be possible to specify specific GPU SKUs when deploying models because the
Kubernetes scheduler treats all GPUs as identical (referring to them all with the generic resources name `nvidia.com/gpu`).

#### Kubernetes Node Discovery Service

This is a requirement for the [NVIDIA GPU Discovery Service for Kubernetes](#nvidia-gpu-discovery-service-for-kubernetes).

#### NVIDIA DCGM Exporter

Provides hardware monitoring and metrics for NVIDIA GPUs and other devices present in the cluster.
Without the metrics this provides, monitoring GPU utilization, temperature and other metrics would not be possible.

While Triton Server has the capability to collect and serve NVIDIA hardware metrics, relying on Triton Server to provide this
service is non-optimal for several reasons.

Firstly, many processes on the same machine querying the NVIDIA device driver for current state, filtering the results for
only values that pertain to the individual process, and serving them via Triton's open-metrics server is as wasteful as the
the number of Triton Server process beyond the first on the node.

> Secondly, due to the need to interface with the kernel-mode driver to retrieve hardware metrics queries get serialized adding

Author: fixed up.

additional overhead and latency to the system.

Finally, the rate at which metrics are collected from Triton Server is not the same as the rate at which metrics are collected
from the DCGM Exporter.
Separating the metrics collection from Triton Server allows for customized metric collection rates, which enables us to
further minimize the process overhead placed on the node.

##### Why is the DCGM Exporter Values File Custom?

I decided to use a custom values file when installing the DCGM Exporter Helm chart for several reasons.

Firstly, it is my professional opinion that every container in a cluster should specify resource limits and requests.
Not doing so opens the node up to a number of difficult to diagnose failure conditions related to resource exhaustion.
Out of memory errors are the most obvious and easiest to root cause.

> Additionally, difficult to reproduce, transient timeout and timing errors caused CPU over subscription can easily happen when

Author: fixed up.

Fortunately, the DCGM Helm chart makes this easy by support node selector options.

Thirdly, because nodes with NVIDIA GPUs have been tainted with the `nvidia.com/gpu=present:NoSchedule` that prevents any
pod which does not explicitly tolerate the taint from be assigned to the node, I need to add the tolerations to the DCGM
Exporter pod.

Finally, the default Helm chart for DCGM Exporter is missing the required `--kubernetes=true` option being passed in via
command line options when the process is started.
Without this option, DCGM Exporter does not correctly associate hardware metrics with the pods actually using it, and
there would be mechanism for understand how each pod uses the GPU resources assigned to it.

#### Prometheus Stack for Kubernetes

Provides metrics collection and aggregation services for the cluster.
While there are other tools capable of providing similar services, we found the Prometheus Stack for Kubernetes was the
easiest to install and configure.
Additionally, the automatic inclusion of a Grafana based user interface made visualization of the cluster's current health
easier to setup.

> Out initial work on this document were based on another metrics service, but we found the configuration of metrics collection

Author: The Kubernetes Metrics server. It just wasn't helpful other than to serve basic metrics to k8s itself.

Prometheus is fairly feature rich, and the ability to create rules was powerful and very useful.

deployed Prometheus server.

#### Why Use the Triton CLI and Not Other Tools Provided by NVIDIA?

I chose to use the new [Triton CLI](https://github.com/triton-inference-server/triton_cli) tool to optimize models for
TensorRT-LLM instead of other available tools for a couple of reasons.

Firstly, using the Triton CLI simplifies the conversion and optimization of models into a single command.

Secondly, relying on the Triton CLI simplifies the creation of the container because all requirements were met with a single
`pip install` command.

##### Why Use a Custom Branch of Triton CLI Instead of an Official Release?

I decided to use a custom [branch of Triton CLI](https://github.com/triton-inference-server/triton_cli/tree/jwyman/aslb-mn)
because there are features this guide needed that were not present in any of the official releases available.
The branch is not a Merge Request because the method used to add the needed features does not aligned with changes the
maintainers have planned.
Once we can achieve alignment, this guide will be updated to use an official release.

### Why Does the Chart Run a Python Script Instead of Triton Server Directly?

There are two reasons:

> 1.  In order to retrieve a model from Hugging Face, convert and optimize it for TensorRT-LLM, and cache it on the host decided

Author: fixed up.

    that [pod initialization container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) was the most
    straightforward solution.

    In order to make the best use of the initialization container I chose to use a custom [server.py](./containers/server.py)
    script that made of the new [Triton CLI](https://github.com/triton-inference-server/triton_cli) tool.

> 2.  Multi-GPU deployments require a rather specialized command line, and generating it using Helm chart scripting was not

Author: I like the suggestion. Used "run" instead of "invocation".

    Leveraging the custom Python script was the logical, and easiest, solution.

#### Why is the Python Written Like That?

Because I'm not a Python developer, but I am learning!
My background is in C/C++ with plenty of experience with shell scripting languages.

### Why Use a Custom Triton Image?

I decided to use a custom image for a few reasons.

1.  Given the answer above and the use of Triton CLI and a custom Python script, the initialization container needed both
    components pre-installed in it to avoid unnecessary use of ephemeral storage.

    _Use of ephemeral storage can lead to pod eviction, and therefore should be avoided whenever possible._

2.  Since the Triton + TRT-LLM image is already incredibly large, I wanted to avoid consuming additional host storage space
    with yet another container image.

    Additionally, the experience of a pod appearing to be stuck in the `Pending` state while it download a container prior to
    the initialization container is easier to understand compared to a short `Pending` state before the initialization
    container, followed by a much longer `Pending` state before the Triton Server can start.

> 3.  I wanted custom, a constant environment variable set for `ENGINE_DEST_PATH` that could be used by both the initialization

Author: fixed up.

    and Triton Server containers.

### What is the `client/` Folder For?

I decided to include the tools I used to validate this guide, and the deployment definitions in the `client/` folder are a key
piece of that effort.
You can use them yourself, if you want to.
All that is required is to run (example for `llama-3-8b`) `kubectl apply -f ./clients/llama-3-8b.yaml` to create the
deployment followed by `kubectl scale deployment/llama-3-8b --replicas=<number_of_desired_clients>`.

As you increase the number of clients generating inference requests for a given Triton Server deployment, load will increase
on the server and the queue-to-compute ratio will eventually cause the horizontal pod autoscaler to increase the number of
Triton Server instancing handle requests until the desired ratio is achieved.

> Decreasing the number fo clients will have the inverse effect and reduce the number of Triton Server instances deployed.

Reviewer 2: ```suggestion
Decreasing the number of clients will have the inverse effect and reduce the number of Triton Server instances deployed.
```

Author: fixed up.

> 

Author: this was done. I've added it to the top of the document.

<!---
# Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
--->

# Autoscaling and Load Balancing Generative AI w/ Triton Server and TensorRT-LLM

> Setting up autoscaling and load balancing using Triton Inference Server, TensorRT-LLM or vLLM, and Kubernetes is not difficult,

Reviewer: ```suggestion
Setting up autoscaling and load balancing for large language models served by Triton Inference Server is not difficult,
```

> but it does require preparation.

Reviewer: ```suggestion
but does require preparation.
```

> This guide aims to help you automated acquisition of models from Hugging Face, minimize time spent optimizing models for

Reviewer: ```suggestion
This guide outlines the steps to download models from Hugging Face, optimize them for TensorRT, and configure automatic scaling and load balancing in Kubernetes.
```

TensorRT, and configuring automatic scaling and load balancing for your models. This guide does not cover Kubernetes'
basics, secure ingress/egress from your cluster to external clients, nor cloud provider interfaces or implementations of
Kubernetes.

> The intent of setting up autoscaling is that it provides the ability for LLM based services to automatically and dynamically

Reviewer: ```suggestion
When configured properly autoscaling enables LLM based services to allocate and deallocate resources automatically based on the current load.
```

adapt to the current workload intensity.

> As the number of clients generating inference requests for a given Triton Server deployment, load will increase on the server

Reviewer: ```suggestion
In this tutorial, as the number of clients grow for a given Triton Server deployment, the inference load on the server increases
```

  * [Custom Container Image](#custom-container-image)
  * [Kubernetes Pull Secrets](#kubernetes-pull-secrets)
* [Triton Deployment](#triton-deployment)
  * [Deploying Single GPU Models](#deploying-single-gpu-models)
  * [Deploying Models Too Large for a Single GPU](#deploying-models-too-large-for-a-single-gpu)
  * [Utilizing Multiple GPU SKUs](#utilizing-multiple-gpu-skus)
  * [Monitoring Triton in Kubernetes](#monitoring-triton-in-kubernetes)
* [Developing this Guide](#developing-this-guide)

Prior to beginning this guide/tutorial you will need a couple of things.

* Kubernetes Control CLI (`kubectl`)
  [ [documentation](https://kubernetes.io/docs/reference/kubectl/introduction/)
  | [download](https://kubernetes.io/releases/download/) ]
* Helm CLI (`helm`)
  [ [documentation](https://helm.sh/)
  | [download](https://helm.sh/docs/intro/install) ]
* Docker CLI (`docker`)
  [ [documentation](https://docs.docker.com/)
  | [download](https://docs.docker.com/get-docker/) ]
* Decent text editing software for editing YAML files.
* Kubernetes cluster.
* Fully configured `kubectl` with administrator permissions to the cluster.

## Cluster Setup

> The following instructions are setting up Horizontal Pod Autoscaling (HPA) for Triton Server in a Kubernetes cluster.

Reviewer: ```suggestion
The following instructions detail how to set up Horizontal Pod Autoscaling (HPA) for Triton Inference Server in a Kubernetes cluster.
```

The following series of steps are intended to prepare a fresh cluster.
For clusters in varying states, it is best to coordinate with your cluster administrator before installing new services and
capabilities.

#### Kubernetes Node Feature Discovery service

1.  Add the Kubernetes Node Feature Discovery chart repository to the local cache.

    ```bash
    helm repo add kube-nfd https://kubernetes-sigs.github.io/node-feature-discovery/charts \
      && helm repo update
    ```

2.  Run the command below to install the service.

    ```bash
    helm install -n kube-system node-feature-discovery kube-nfd/node-feature-discovery \
      --set nameOverride=node-feature-discovery \
      --set worker.tolerations[0].key=nvidia.com/gpu \
      --set worker.tolerations[0].operator=Exists \
      --set worker.tolerations[0].effect=NoSchedule
    ```

    > [!Note]
    > The above command sets toleration values which allow for the deployment of a pod onto a node with
    > a matching taint.
    > See this document's [prerequisites](#prerequisites) for the taints this document expected to have been applied to GPU
    > nodes in the cluster.

#### NVIDIA Device Plugin for Kubernetes

> 1.  This step is unnecessary if the Device Plugin has already been installed in your cluster.

Reviewer: ```suggestion
1.  This step is not needed if the Device Plugin has already been installed in your cluster.
```

>     Cloud provider turnkey Kubernetes clusters, such as those from AKS, EKS, and GKE, often have the Device Plugin

Reviewer: ```suggestion
    Cloud providers with turnkey Kubernetes clusters, such as those from AKS, EKS, and GKE, often install the Device Plugin automatically when a GPU node is added to the cluster. 
```

    To check if your cluster requires the NVIDIA Device Plugin for Kubernetes, run the following command and inspect
    the output for `nvidia-device-plugin-daemonset`.

    ```bash
    kubectl get daemonsets --all-namespaces
    ```

    Example output:
    ```text
    NAME                                          DESIRED  CURRENT  READY  UP-TO-DATE  AVAILABLE
    kube-proxy                                    6        6        6      6           6
    kube-system   node-feature-discovery-worker   1        1        1      1           1
    nvidia-device-plugin-daemonset                6        6        6      6           6
    ```

2.  If `nvidia-device-plugin-daemonset` is not listed, run the command below to install the plugin.
    Once installed it will provide containers access to GPUs in your clusters.

    For additional information, see
    [Github/NVIDIA/k8s-device-plugin](https://github.com/NVIDIA/k8s-device-plugin/blob/main/README.md).

    ```bash
    kubectl create -f https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/v0.15.0/deployments/static/nvidia-device-plugin.yml
    ```

#### NVIDIA GPU Feature Discovery Service

> 1.  This step is unnecessary if the service has already been installed in your cluster.

Reviewer: ```suggestion
1.  This step is not needed if the service has already been installed in your cluster.
```

>     To check if your cluster requires the NVIDIA Device Plugin for Kubernetes, run the following command and inspect

Reviewer: ```suggestion
    To check if your cluster requires the NVIDIA GPU Feature Discovery Service, run the following command and inspect
```

    kube-system   node-feature-discovery-worker         6         6         6       6            6
    kube-system   nvidia-device-plugin-daemonset        6         6         6       6            6
    ```

2.  If `gpu-feature-discover` is listed, skip this step and the next.

    Otherwise, use the YAML file below to install the GPU Feature Discovery service.

    > [nvidia_gpu-feature-discovery_daemonset.yaml](nvidia_gpu-feature-discovery_daemonset.yaml)

    The file above was created by downloading its contents from
    [GitHub/NVIDIA](https://raw.githubusercontent.com/NVIDIA/gpu-feature-discovery/v0.8.2/deployments/static/gpu-feature-discovery-daemonset.yaml)
    and modified specifically for Triton Server autoscaling.

    ```bash
    curl https://raw.githubusercontent.com/NVIDIA/gpu-feature-discovery/v0.8.2/deployments/static/gpu-feature-discovery-daemonset.yaml \
      >  nvidia_gpu-feature-discovery_daemonset.yaml
    ```

3.  Then run the command below to install the

    ```bash
    kubectl apply -f ./nvidia_gpu-feature-discovery_daemonset.yaml
    ```

### Metrics Collection Services

> Your cluster is now up, running, and can even assign GPU resources to containers.

Reviewer: ```suggestion
Your cluster is now up, running, and can assign GPU resources to containers.
```

Reviewer: should containers be `pods` ?

Next, we have to setup metrics collection for DCGM and Triton Server.

> Metrics provide insight to the Kubernetes Horizontal Pod Autoscaler service and enable it to make autoscaling decisions based

Reviewer: ```suggestion
Metrics services provide utilization and availability data to the Kubernetes Horizontal Pod Autoscaler to be used to make autoscaling decisions.
```

1.  Add the Prometheus Community chart repository to the local cache.

    ```bash
    helm repo add prometheus-community https://prometheus-community.github.io/helm-charts \
      && helm repo update
    ```

2.  Run the command below to install the Prometheus Kubernetes Stack Helm chart.

    ```bash
    helm install -n monitoring prometheus prometheus-community/kube-prometheus-stack \
      --set tolerations[0].key=nvidia.com/gpu \
      --set tolerations[0].operator=Exists \
      --set tolerations[0].effect=NoSchedule
    ```

    > [!Note]
    > The above command sets toleration values which allow for the deployment of a pod onto a node with
    > a matching taint.
    > See this document's [prerequisites](#prerequisites) for the taints this document expected to have been applied to GPU
    > nodes in the cluster.

#### NVIDIA Data Center GPU Manager (DCGM) Exporter

The best solution for management of GPUs in your cluster is
[NVIDIA DCGM](https://docs.nvidia.com/data-center-gpu-manager-dcgm)(DCGM).
However, for this example we do not need the entirety of the DCGM stack.

> Instead, we'll use the steps below to install the just [DCGM Exporter](https://github.com/NVIDIA/dcgm-exporter) to enable the

Reviewer: ```suggestion
Instead, we'll use the steps below to install just the [DCGM Exporter](https://github.com/NVIDIA/dcgm-exporter) to enable the
```
